Expose raw saved tensors for custom functions

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60975
* #60663
* #60565
* #60685
* **#60551**

Summary: This PR is part of a stack aimed to address #58512. 
Provide bindings for the `SavedVariable` class (for custom functions only). The `SavedTensor` object can not be created from Python and currently does not expose any method.
Undefined behavior can arise if the user decides to keep a reference to this `SavedTensor` after the underlying `SavedVariable` has been freed.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466228](https://our.internmc.facebook.com/intern/diff/D29466228)

Varal7 committed:
Expose raw saved tensors for custom functions

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -743,6 +743,25 @@ PyObject *THPFunction_saved_variables(THPFunction *self, void *_unused)
+  if (!saved)
+    return nullptr;
+  for(const auto i : c10::irange(num_saved)) {
+    py::object obj = py::cast(saved_variables[i], py::return_value_policy::reference);
```
This should probably be abstracted away into another function

Varal7 replied:
@soulitzer This snippet code is used three times but not sure where the best place to put it would be

soulitzer replied:
Where else is it being used?

soulitzer replied:
From here https://pybind11.readthedocs.io/en/stable/advanced/functions.html using py::return_value_policy::reference might lead to undefined behavior if the c++ side is freed and we keep using it on the Python side

Varal7 replied:
They are used [here](https://github.com/pytorch/pytorch/pull/60565/files#diff-2bf4105c8354f4a2656b421e0dc44d72542efdde41070ee67a81c7d99c34939aR216) and [here](https://github.com/pytorch/pytorch/pull/60565/files#diff-2bf4105c8354f4a2656b421e0dc44d72542efdde41070ee67a81c7d99c34939aR201)

Varal7 commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -743,6 +743,25 @@ PyObject *THPFunction_saved_variables(THPFunction *self, void *_unused)
 
+PyObject *THPFunction_raw_saved_tensors(THPFunction *self, void *_unused)
+{
+  HANDLE_TH_ERRORS
```
Maybe refactor this and unpack_saved_variables?

Varal7 replied:
The function I meant to refactor was ihttps://github.com/pytorch/pytorch/blob/7e619b95880cdec72d1353c9c99dedbc2575ad36/torch/csrc/autograd/python_function.cpp#L689

(they already slightly diverged, but I don't have strong preference in favor of refactoring)

Varal7 commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -743,6 +743,25 @@ PyObject *THPFunction_saved_variables(THPFunction *self, void *_unused)
+  END_HANDLE_TH_ERRORS
+}
+
+
```
Extra line

Varal7 committed:
Update on "Expose raw saved tensors for custom functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -845,6 +864,7 @@ PyObject* getRequiresGrad(PyObject* obj, void* _unused) {
 static struct PyGetSetDef THPFunction_properties[] = {
   {"saved_tensors", (getter)THPFunction_saved_tensors, nullptr, nullptr, nullptr},
   {"saved_variables", (getter)THPFunction_saved_variables, nullptr, nullptr, nullptr},
+  {"raw_saved_tensors", (getter)THPFunction_raw_saved_tensors, nullptr, nullptr, nullptr},
```
Maybe add a small test to make sure we can access this from `grad_fn.raw_saved_tensor` and test what happens if there aren't any saved etc.

albanD replied:
Why is this public API? Shouldn't it be `_raw_saved_tensors` ?

Varal7 replied:
I used `raw_saved_tensors` for consistency with the implicit rule "prefix `saved` with `raw_` and you get the raw version". Would you like me to change it?

soulitzer reviewed (commented):
Looks ok, the main issue I think is how to deal with the case when backward already frees the SavedVariable. Maybe theres some way to do this in pybind, but if we use the plain old PyObject struct, we can have it hold a weak_ptr and be able to raise an error if c++ side is already freed

Varal7 commented:
> Looks ok, the main issue I think is how to deal with the case when backward already frees the SavedVariable. Maybe theres some way to do this in pybind, but if we use the plain old PyObject struct, we can have it hold a weak_ptr and be able to raise an error if c++ side is already freed

Thanks for the review! 
Yes the behavior after #60565 is that accessing the SavedVariable will raise an error (https://github.com/pytorch/pytorch/pull/60565/files#diff-1fdfd061a92957fd11edbaceb538f7186f185b9fb22b7219107b007f7df557fbR750)
The error could be a tiny bit more explicit (in this case we know the user is explicitly trying to access the saved variable, and not trying to backward, but the current error message already mentions that it could be they are trying to access the saved variable).

I can rebase to include it in this PR.

soulitzer commented:
Np!
Ahh yes, that error is probably good to include as well in this PR for completeness.
The situation I was thinking about is slightly different, i.e., if the user calls raw_saved_tensors and saves a reference to a SavedVariable, then calls backward.

albanD commented:
Removing myself as @soulitzer is handling this one.
You can add me back if you have a specific question for me!

Varal7 committed:
Update on "Expose raw saved tensors for custom functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 commented:
> Np!
> Ahh yes, that error is probably good to include as well in this PR for completeness.
> The situation I was thinking about is slightly different, i.e., if the user calls raw_saved_tensors and saves a reference to a SavedVariable, then calls backward.

Good point. My proposal is the following.
Because we only (and only plan to) expose the method `SavedVariable::register_hooks`, we can check there that the variable has not already been freed. 
So if the user keeps a reference to the raw_saved_tensors after he has called backward he will just hold a useless object and will have a nice error message if he tries to call register_hook on it.
Like so: https://github.com/pytorch/pytorch/pull/60663/files#diff-cc9fba479b5beae06b2eea2e390d17796e0341c5b037a20b5bcaccbb0c341030R156

soulitzer commented:
That's a good plan - but there are also some cases where the entire SavedVariable is destroyed no (not just its data field)? 
For example all custom functions (we could add a test for this here), and some codegen'd functions store saved variable in a vector, and on `release_variables`, the vector is cleared.

Varal7 commented:
While I do agree with you, I've written tests to test those cases ([here](https://github.com/pytorch/pytorch/pull/60685/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR4801) and [here](https://github.com/pytorch/pytorch/pull/60565/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR4686)).

What is confusing to me is that the following code does trigger `RuntimeError: Calling register_hook on a saved tensor after it has been freed.` as we would want.

```python
class MyFn(Function):
    @staticmethod
    def forward(ctx, x):
        ctx.save_for_backward(x)
        return x

    @staticmethod
    def backward(ctx, g):
        return g, None

y = MyFn.apply(torch.ones(5, requires_grad=True))
t = y.grad_fn.raw_saved_tensors
y.sum().backward() # calls grad_fn->release_variables() and the destructors for each SavedVariable (but not `.reset()`)
t[0].register_hooks(lambda x: x, lambda x: x)

soulitzer commented:
It could be because vector.clear just updates its size, but capacity is required to stays the same, and thus the memory is still reserved by the vector.
Maybe we should tell the vector `shrink_to_fit`, to actually release the memory (though technically that is still not guaranteed).
Could be interesting as well to see if this improves memory usage too

Varal7 committed:
Update on "Expose raw saved tensors for custom functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Expose raw saved tensors for custom functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

soulitzer reviewed (approved):
LGTM

soulitzer commented on torch/csrc/autograd/init.cpp:
```diff
@@ -263,6 +264,11 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
 
+  py::class_<torch::autograd::SavedVariable>(m, "SavedTensor")
+    .def(py::init([]()->torch::autograd::SavedVariable {
+      TORCH_CHECK(false, "Trying to create a SavedTensor object from Python is forbidden.");
```
Also add a test for this?

albanD commented on test/test_autograd.py:
```diff
@@ -4758,6 +4758,44 @@ def test_grad_fn_attr_bindings(self):
+
+                @staticmethod
+                def backward(ctx, g):
+                    return g, None
```
Why is there two returns here while you have a single input for the forward?

Note to self: I still don't understand why we allow trailing None here :p

albanD reviewed (commented):
Just a couple of small comments.

Varal7 committed:
Update on "Expose raw saved tensors for custom functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466228).

Varal7 committed:
Update on "Expose raw saved tensors for custom functions"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466228](https://our.internmc.facebook.com/intern/diff/D29466228)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466228).
